crates/uv-auth/src/middleware.rs

```diff
@@ -100,88 +140,199 @@ impl Middleware for AuthMiddleware {
+            let credentials = self.cache().get_realm(
+                Realm::from(retry_request.url()),
+                credentials
+                    .map(|credentials| credentials.to_username())
+                    .unwrap_or(Username::none()),
+            );
+
+            if let Some(credentials) = credentials.as_ref() {
+                if credentials.password().is_some() {
+                    trace!(
+                        "Retrying request for {url} with credentials from cache {credentials:?}"
+                    );
+                    retry_request = credentials.authenticate(retry_request);
+                    return self
+                        .complete_request(None, retry_request, extensions, next)
+                        .await;
+                }
+            }
+
+            if let Some(credentials) = self
+                .fetch_credentials(credentials.as_deref(), retry_request.url())
+                .await
+            {
+                retry_request = credentials.authenticate(retry_request);
+                trace!("Retrying request for {url} with {credentials:?}");
+                return self
+                    .complete_request(Some(Arc::new(credentials)), retry_request, extensions, next)
+                    .await;
+            } else {
+                if let Some(credentials) = credentials.as_ref() {
+                    if !attempt_has_username {
+                        trace!(
+                            "Retrying request for {url} with username from cache {credentials:?}"
+                        );
+                        retry_request = credentials.authenticate(retry_request);
+                        return self
+                            .complete_request(None, retry_request, extensions, next)
+                            .await;
+                    }
+                }
+
+                Ok(response)
+            }
+        } else {
+            Ok(response)
+        }
+    }
+}
+
+impl AuthMiddleware {
+    /// Run a request to completion.
+    ///
+    /// If credentials are present, insert them into the cache on success.
+    async fn complete_request(
+        &self,
+        credentials: Option<Arc<Credentials>>,
+        request: Request,
+        extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        if let Some(credentials) = credentials {
+            let url = request.url().clone();
+            let result = next.run(request, extensions).await;
+
+            // Update the cache with new credentials on a successful request
+            if result
+                .as_ref()
+                .is_ok_and(|response| response.error_for_status_ref().is_ok())
+            {
+                trace!("Updating cached credentials for {url} to {credentials:?}");
+                self.cache().insert(&url, credentials)
+            };
+
+            result
+        } else {
+            next.run(request, extensions).await
+        }
```

Comment:
Continuing on my early return crusade, I'd write this as:

```rust
let Some(credentials) = credentials else {
    return next.run(request, extensions).await;
};
let url = request.url().clone();
// and so on...
```

Reply:
Ah I forget about `let else`